src/Servers/Kestrel/Core/src/IHttpsConfigurationService.cs

```diff
@@ -0,0 +1,81 @@
+/// </summary>
+internal interface IHttpsConfigurationService
+{
+    /// <summary>
+    /// If this property returns false, then methods other than <see cref="Initialize"/> will throw.
+    /// The most obvious way to make this true is to call <see cref="Initialize"/>, but some implementations
+    /// may offer alternative mechanisms.
+    /// </summary>
+    bool IsInitialized { get; }
+
+    /// <summary>
+    /// Replaces the implementations off all other methods with functioning (as opposed to throwing) versions.
+    /// </summary>
+    void Initialize(
+        IHostEnvironment hostEnvironment,
+        ILogger<KestrelServer> serverLogger,
+        ILogger<HttpsConnectionMiddleware> httpsLogger);
+
+    #region For KestrelConfigurationLoader
+    /// <summary>
+    /// Applies various configuration settings to <paramref name="httpsOptions"/> and <paramref name="endpoint"/>.
+    /// </summary>
+    void ApplyHttpsConfiguration(
+        HttpsConnectionAdapterOptions httpsOptions,
+        EndpointConfig endpoint,
+        KestrelServerOptions serverOptions,
+        CertificateConfig? defaultCertificateConfig,
+        ConfigurationReader configurationReader);
+
+    /// <summary>
+    /// Calls an appropriate overload of <see cref="Microsoft.AspNetCore.Hosting.ListenOptionsHttpsExtensions.UseHttps(ListenOptions)"/>
+    /// on <paramref name="listenOptions"/>, with or without SNI, according to how <paramref name="endpoint"/> is configured.
+    /// </summary>
+    /// <returns>Updated <see cref="ListenOptions"/> for convenient chaining.</returns>
+    ListenOptions UseHttpsWithSni(ListenOptions listenOptions, HttpsConnectionAdapterOptions httpsOptions, EndpointConfig endpoint);
+
+    /// <summary>
+    /// Retrieves the default or, failing that, developer certificate from <paramref name="configurationReader"/>.
+    /// </summary>
+    CertificateAndConfig? LoadDefaultCertificate(ConfigurationReader configurationReader);
+    #endregion
+
+    #region For TransportManager
+    /// <summary>
+    /// Updates <paramref name="features"/> with multiplexed transport (i.e. HTTP/3) features based on
+    /// the configuration of <paramref name="listenOptions"/>.
+    /// </summary>
+    void PopulateMultiplexedTransportFeatures(FeatureCollection features, ListenOptions listenOptions);
+    #endregion
+
+    #region For AddressBinder
+    /// <summary>
+    /// Calls <see cref="Microsoft.AspNetCore.Hosting.ListenOptionsHttpsExtensions.UseHttps(ListenOptions)"/>
+    /// on <paramref name="listenOptions"/>.
+    /// </summary>
+    /// <returns>Updated <see cref="ListenOptions"/> for convenient chaining.</returns>
+    ListenOptions UseHttpsWithDefaults(ListenOptions listenOptions);
+    #endregion
+}
+
+/// <summary>
+/// A <see cref="Certificate"/>-<see cref="CertificateConfig"/> pair.
+/// </summary>
+internal record CertificateAndConfig(X509Certificate2 Certificate, CertificateConfig CertificateConfig);
```

Comment:
We should avoid the blot of records in the codebase.

Reply:
+1. More info https://github.com/dotnet/aspnetcore/issues/45859

Reply:
Now a `readonly struct`.